### PR TITLE
Reapply "Use system installed yaml-cpp 0.6 if available (#8)"

### PR DESCRIPTION
**Pull request by ivanpauno**
Reapplies https://github.com/ros2/yaml_cpp_vendor/pull/8.

I will try to see why it was failing on macOS, by adding some extra debug messages.

### CODE REVIEW

**Comment by ivanpauno** [comment]
@seanyen FYI

**Comment by ivanpauno** [comment]
CI with the new debug message: https://ci.ros2.org/job/ci_osx/9833/.

**Comment by ivanpauno** [comment]
@seanyen there's something quite weird here, this seems to be finding a previously downloaded version of `yaml-cpp` in the same workspace:

```
Found yaml-cpp 0.6.2 in path /Users/osrf/jenkins-agent/workspace/nightly_osx_extra_rmw_release/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build/yaml-cpp-config.cmake
```

Maybe we're not cleaning up the workspace in our macOS machines between builds, that would be pretty weird.

Maybe there's a way to call `find_package` that avoid this ... (`NO_CMAKE_BUILDS_PATH` ?)

**Comment by seanyen** [comment]
@ivanpauno Hi, is it possible to turn on [`CMAKE_FIND_DEBUG_MODE`](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_DEBUG_MODE.html)? I think it would be helpful to understand the CMake search order (in macOS).

**Comment by ivanpauno** [comment]
> @ivanpauno Hi, is it possible to turn on [`CMAKE_FIND_DEBUG_MODE`](https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_DEBUG_MODE.html)? I think it would be helpful to understand the CMake search order (in macOS).

Enabled in https://github.com/ros2/yaml_cpp_vendor/pull/16/commits/46ede7b3f01b326b624b8e097b1165d01f8e7583.
new CI: https://ci.ros2.org/job/ci_osx/9840/

---
Maybe this behavior is system independent, and we're only running into the bug on macOS because of not pruning the old workspace.
I imagine this can be tested in the following way:
- Build without system yaml
- Modify something in the CMakelists.txt (e.g. add a debug message), so that cmake does something in next build
- Check if find_package founds the previously downloaded version or not

**Comment by ivanpauno** [comment]
<details>
  <summary>find_package debug output</summary>

```
17:54:29   find_package considered the following paths for yaml-cpp.cmake
17:54:29 
17:54:29     /usr/local/Cellar/cmake/3.17.2/share/cmake/Modules/Findyaml-cpp.cmake
17:54:29 
17:54:29   The file was not found.
17:54:29 
17:54:29   <PackageName>_ROOT CMake variable [CMAKE_FIND_USE_PACKAGE_ROOT_PATH].
17:54:29 
17:54:29     none
17:54:29 
17:54:29   CMAKE_PREFIX_PATH variable [CMAKE_FIND_USE_CMAKE_PATH].
17:54:29 
17:54:29     none
17:54:29 
17:54:29   CMAKE_FRAMEWORK_PATH and CMAKE_APPBUNDLE_PATH variables
17:54:29   [CMAKE_FIND_USE_CMAKE_PATH].
17:54:29 
17:54:29     none
17:54:29 
17:54:29   Env variable yaml-cpp_DIR [CMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH].
17:54:29 
17:54:29     none
17:54:29 
17:54:29   CMAKE_PREFIX_PATH env variable [CMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH].
17:54:29 
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_test
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_target_dependencies
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_targets
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_interfaces
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_dependencies
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_version
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_python
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_libraries
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_include_directories
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_link_flags
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_libraries
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_include_directories
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_definitions
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_core
17:54:29     /usr/local/opt/qt
17:54:29 
17:54:29   
17:54:29 
17:54:29   CMAKE_FRAMEWORK_PATH and CMAKE_APPBUNDLE_PATH env variables
17:54:29   [CMAKE_FIND_USE_CMAKE_ENVIRONMENT_PATH].
17:54:29 
17:54:29   
17:54:29 
17:54:29   Paths specified by the find_package HINTS option.
17:54:29 
17:54:29     none
17:54:29 
17:54:29   Standard system environment variables
17:54:29   [CMAKE_FIND_USE_SYSTEM_ENVIRONMENT_PATH].
17:54:29 
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/venv
17:54:29     /usr/local/opt/ccache/libexec
17:54:29     /usr/local
17:54:29     /usr
17:54:29     /
17:54:29 
17:54:29   
17:54:29 
17:54:29   CMake User Package Registry [CMAKE_FIND_USE_PACKAGE_REGISTRY].
17:54:29 
17:54:29     /Users/osrf/jenkins-agent/workspace/packaging_osx/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build
17:54:29 
17:54:29   
17:54:29 
17:54:29   CMake variables defined in the Platform file
17:54:29   [CMAKE_FIND_USE_CMAKE_SYSTEM_PATH].
17:54:29 
17:54:29     /usr/local/Cellar/cmake/3.17.2
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/yaml_cpp_vendor
17:54:29     /usr/X11R6
17:54:29     /usr/pkg
17:54:29     /opt
17:54:29     /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/usr
17:54:29     /sw
17:54:29     /opt/local
17:54:29     /Users/osrf/Library/Frameworks
17:54:29     /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/Library/Frameworks
17:54:29     /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/Network/Library/Frameworks
17:54:29     /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk/System/Library/Frameworks
17:54:29     /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/Library/Frameworks
17:54:29     /Applications/Xcode.app/Contents/Developer/Library/Frameworks
17:54:29     /Library/Frameworks
17:54:29     /Network/Library/Frameworks
17:54:29     /System/Library/Frameworks
17:54:29     /Applications
17:54:29     /Applications/Xcode.app/Contents/Applications
17:54:29     /Applications/Xcode.app/Contents/Developer/Applications
17:54:29 
17:54:29   
17:54:29 
17:54:29   CMake System Package Registry
17:54:29   [CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY].
17:54:29 
17:54:29     none
17:54:29 
17:54:29   Paths specified by the find_package PATHS option.
17:54:29 
17:54:29     none
17:54:29 
17:54:29   find_package considered the following locations for the Config module:
17:54:29 
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_test/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_test/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_target_dependencies/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_target_dependencies/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_targets/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_targets/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_interfaces/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_interfaces/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_dependencies/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_dependencies/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_version/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_version/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_python/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_python/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_libraries/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_libraries/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_include_directories/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_include_directories/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_link_flags/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_link_flags/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_libraries/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_libraries/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_include_directories/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_include_directories/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_definitions/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_export_definitions/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_core/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/ws/install/ament_cmake_core/yaml-cpp-config.cmake
17:54:29     /usr/local/opt/qt/yaml-cppConfig.cmake
17:54:29     /usr/local/opt/qt/yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/venv/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/ci_osx/venv/yaml-cpp-config.cmake
17:54:29     /usr/local/opt/ccache/libexec/yaml-cppConfig.cmake
17:54:29     /usr/local/opt/ccache/libexec/yaml-cpp-config.cmake
17:54:29     /usr/local/yaml-cppConfig.cmake
17:54:29     /usr/local/yaml-cpp-config.cmake
17:54:29     /usr/yaml-cppConfig.cmake
17:54:29     /usr/yaml-cpp-config.cmake
17:54:29     /yaml-cppConfig.cmake
17:54:29     /yaml-cpp-config.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/packaging_osx/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build/yaml-cppConfig.cmake
17:54:29     /Users/osrf/jenkins-agent/workspace/packaging_osx/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build/yaml-cpp-config.cmake
17:54:29 
17:54:29   The file was found at
17:54:29 
17:54:29     /Users/osrf/jenkins-agent/workspace/packaging_osx/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build/yaml-cpp-config.cmake
17:54:29 
17:54:29 
17:54:29 
17:54:29 -- Found yaml-cpp 0.6.2 in path /Users/osrf/jenkins-agent/workspace/packaging_osx/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build/yaml-cpp-config.cmake
```
</details>

@seanyen FYI

**Comment by seanyen** [comment]
@ivanpauno Thanks for the logs. Ok, the CMake config files are discovered from [`CMake User Package Registry`](https://cmake.org/cmake/help/v3.18/manual/cmake-packages.7.html#user-package-registry).

```
CMake User Package Registry [CMAKE_FIND_USE_PACKAGE_REGISTRY].
/Users/osrf/jenkins-agent/workspace/packaging_osx/ws/build/yaml_cpp_vendor/yaml_cpp-0f9a586-prefix/src/yaml_cpp-0f9a586-build
```

There is a `NO_CMAKE_PACKAGE_REGISTRY` option to skip the discovery, but I think for CI service, it makes more sense to nuke `~/.cmake/packages/` for reproducible builds.

**Comment by ivanpauno** [comment]
> There is a NO_CMAKE_PACKAGE_REGISTRY option to skip the discovery, but I think for CI service, it makes more sense to nuke ~/.cmake/packages/ for reproducible builds.

Doing that would make builds more reproducible, but we still need `NO_CMAKE_PACKAGE_REGISTRY` (and maybe some of the other restrictions).
Without that, workspace A can find the library downloaded in workspace B, and people will run into this same issue.

That could be fixed by always installing the environment hook, but I think it's better to only look for the "system installed" yaml-cpp version.

I will reread find_package docs and commit a fix.

**Comment by ivanpauno** [comment]
New CI, testing `rosbag2_tests`:
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9846/)

@seanyen does this make sense to you?

File: CMakeLists.txt
Comment (dirk-thomas) [inline CMakeLists.txt:69]: What if workspace B is an underlay of workspace A? Wouldn't it be desired to use it from workspace A then?

Comment (ivanpauno) [reply CMakeLists.txt:69]: I don't see why both overlay and underlay would have a copy of `yaml_cpp_vendor`.
I'm not sure if that case can be handled separately.

The idea is to force cmake to look for the system installed version, and avoid the issue previously experienced on CI.

Comment (dirk-thomas) [reply CMakeLists.txt:69]: > I don't see why both overlay and underlay would have a copy of `yaml_cpp_vendor`.

A Docker image contains a workspace building ROS 2 from source  - including `yaml_cpp_vendor`. Afterwards a user creates a workspace on top and also places `yaml_cpp_vendor` in it - either in order to get something newer or just because even though it isn't necessary.

**Comment by seanyen** [comment]
> Doing that would make builds more reproducible, but we still need `NO_CMAKE_PACKAGE_REGISTRY` (and maybe some of the other restrictions).
> Without that, workspace A can find the library downloaded in workspace B, and people will run into this same issue.

I don't have any objections adding `NO_CMAKE_PACKAGE_REGISTRY` (Personally I am not a fan for CMake package registry). As a user building ROS from sources heavily, I think it would be helpful to mention this problem in the [ROS2 docs](https://index.ros.org/doc/ros2/Installation/Foxy/) and point people to [`Disabling the Package Registry`](https://cmake.org/cmake/help/v3.1/manual/cmake-packages.7.html#disabling-the-package-registry), in case he/she wants a isolated build.

File: CMakeLists.txt
Comment (seanyen) [reply CMakeLists.txt:69]: > A Docker image contains a workspace building ROS 2 from source - including `yaml_cpp_vendor`. Afterwards a user creates a workspace on top and also places `yaml_cpp_vendor` in it - either in order to get something newer or just because even though it isn't necessary.

I think this case is covered by `colcon\ament_cmake` doing great job to properly order `CMAKE_PREFIX_PATH`? 😉

One caveat of the package registry is that the search order is [undefined](https://gitlab.kitware.com/cmake/community/-/wikis/doc/tutorials/Package-Registry):

```
The search order among package registry entries for a single package is unspecified and the entry names (hashes in this example) have no meaning.
```

In turns, I think there is no workspace overlay concept for CMake package registry?

Comment (ivanpauno) [reply CMakeLists.txt:69]: > A Docker image contains a workspace building ROS 2 from source - including yaml_cpp_vendor. Afterwards a user creates a workspace on top and also places yaml_cpp_vendor in it - either in order to get something newer or just because even though it isn't necessary.

In that case, yes it will be downloaded twice.
But I don't see how I can detect that particular use case.

Using the library downloaded in another workspace can be quite confusing (if there's no underlay/overlay relation).
e.g.: build workspace A, build ws B, remove A, B now has runtime problems for finding yaml-cpp.

If you have any suggestion on how to solve that issue I'm willing to apply them, but I can't think of a way of handling that case correctly.
This PR is still an improvement over the current master, which always downloads yaml-cpp.

Comment (dirk-thomas) [reply CMakeLists.txt:69]: > In that case, yes it will be downloaded twice.
> But I don't see how I can detect that particular use case.

A vendor package does either of the following two options:
* always build the vendored code
* check if the vendored code is already findable / usable (maybe in a new enough version) and only build it if that is not already the case

Based on that I think the vendor package in the overlay should be able to detect either a system installed version or itself in an underlay workspace. I don't see why that shouldn't be possible.

> Using the library downloaded in another workspace can be quite confusing (if there's no underlay/overlay relation).

If there is no underlay/overlay relation why would it find the library in that other workspace?

> build workspace A, build ws B, remove A, B now has runtime problems for finding yaml-cpp.

This is the case for any kind of package and has nothing to do with the CMake registry. Even when you do the same in a single workspace (build pkg C which depends on D which is also in the workspace, then delete D and try rebuilding).

Comment (ivanpauno) [reply CMakeLists.txt:69]: > If there is no underlay/overlay relation why would it find the library in that other workspace?

AFAIU there's only one package registry per user: https://cmake.org/cmake/help/v3.2/manual/cmake-packages.7.html#user-package-registry.

> This is the case for any kind of package and has nothing to do with the CMake registry. Even when you do the same in a single workspace (build pkg C which depends on D which is also in the workspace, then delete D and try rebuilding).

Sure, but I expect that to fail.
If `yaml_cpp_vendor` found a `yaml-cpp` library in a unrelated workspace through the cmake package registry, that would be more confusing.

Comment (dirk-thomas) [reply CMakeLists.txt:69]: We should maybe consider a more generic approach like colcon/colcon-cmake#91 rather then a one-off patch in this repo.

Comment (ivanpauno) [reply CMakeLists.txt:69]: > We should maybe consider a more generic approach like [colcon/colcon-cmake#91](https://github.com/colcon/colcon-cmake/issues/91) rather then a one-off patch in this repo.

The linked example is exactly the same problem that happens here, a vendor package doing find_package quiet and then downloading/building if not found.
IMO, the problem is that the vendor package is not well implemented, not a missing feature in colcon/ament.

According to [cmake docs](https://cmake.org/cmake/help/v3.1/variable/CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY.html):

> In some cases, for example to locate only system wide installations, it is not desirable to use the User Package Registry when searching for packages

`NO_CMAKE_PACKAGE_REGISTRY` is exactly intended to solve this issue.
Even if we want to globally disable the package registry in colcon by default, the vendor package should be using this option IMO.

> I think this case is covered by colcon\ament_cmake doing great job to properly order CMAKE_PREFIX_PATH? wink

I think that's true. In the case of an overlay/underlay, the library isn't going to be downloaded twice.

**Comment by seanyen** [comment]
@ivanpauno @dirk-thomas Just a friendly ping. Do we have any plans for the macOS build breaks?

**Comment by ivanpauno** [comment]
> @ivanpauno @dirk-thomas Just a friendly ping. Do we have any plans for the macOS build breaks?

I think the current patch is fine, as it is an improvement over how it used to work and it doesn't introduce any new problems.
@dirk-thomas can give you feedback of what's missing.

**Comment by dirk-thomas** [comment]
I don't see how applying this patch to this single package can be the right thing to do. Imo it either needs to be applied globally or to all affected packages individually (in this case this patch if fine but the other packages needing the same flag should be identified as part of this ticket).

**Comment by ivanpauno** [comment]
> I don't see how applying this patch to this single package can be the right thing to do

The focus of the PR isn't adding the usage of `NO_CMAKE_PACKAGE_REGISTRY`, it is to avoid building `yaml_cpp_vendor` if you already have a system installed version or if you have already downloaded it in an underlay workspace.
IMHO we shouldn't block a PR because of an unrelated detail.

> Imo it either needs to be applied globally or to all affected packages individually

Though we could force colcon to use `NO_CMAKE_PACKAGE_REGISTRY` globally by default, there might be some people that want to use the package registry. Adding `NO_CMAKE_PACKAGE_REGISTRY` in the `find_package` calls where using the package registry is a problem seems appropriate to me.
Particularly, I think that all vendor packages using the "try find_package, build if it doesn't succeed" approach should be using that option.

Here's a complete list of packages that would need a similar change:
- [libcurl_vendor](https://github.com/ros/resource_retriever/blob/8053dd51632fae012a65f6f704d29d55ae5d856f/libcurl_vendor/CMakeLists.txt#L68)
- [sqlite3_vendor](https://github.com/ros2/rosbag2/blob/f8f5f669270081f9f5969c301b8b2ca39b75abbe/sqlite3_vendor/CMakeLists.txt#L12)
- [zstd_vendor](https://github.com/ros2/rosbag2/blob/f8f5f669270081f9f5969c301b8b2ca39b75abbe/zstd_vendor/CMakeLists.txt#L12)
- [tinyxml_vendor](https://github.com/ros2/tinyxml_vendor/blob/91ad3796eda73cf8cac09f7d0e7b2344d2e9acea/CMakeLists.txt#L12)
- [rviz_assimp_vendor](https://github.com/ros2/rviz/blob/40e3b9ae9712ce4df7ea368e714dbc509064e8b0/rviz_assimp_vendor/CMakeLists.txt#L61)
- [spdlog_vendor](https://github.com/ros2/spdlog_vendor/blob/0223d4586fff5863e6d380c38bde5fee46802af7/CMakeLists.txt#L12)
- [console_bridge_vendor](https://github.com/ros2/console_bridge_vendor/blob/6cffe405a707a4cbf6ff7a1d13ac22552b3e22ae/CMakeLists.txt#L12)
- [google_benchmark_vendor](https://github.com/ament/google_benchmark_vendor/blob/2bfd73afe808ba1ee8be3b87e5eebf4e26d7e759/CMakeLists.txt#L9)
- [eProsima/foonathan_memory_vendor](https://github.com/eProsima/foonathan_memory_vendor/blob/7ca1e109c2d4d7f0b3b03f6880c770ef00ed79de/CMakeLists.txt#L4)

All other vendor packages aren't using the "try find_package then build" approach.

I will create a ticket with this list in `ros2/ros2` after this gets merged.

**Comment by dirk-thomas** [comment]
> IMHO we shouldn't block a PR because of an unrelated detail.

I wouldn't call this an "unrelated detail" if the whole point about `NO_CMAKE_PACKAGE_REGISTRY` is that without it this patch doesn't work as intended since it shows cross talk between unrelated workspaces.

I just raised the concern that the actual question how we should handle this hasn't been answered yet. Please feel free to merge before that is the case if you think that is fine.

**Comment by ivanpauno** [comment]
> I wouldn't call this an "unrelated detail" if the whole point about NO_CMAKE_PACKAGE_REGISTRY is that without it this patch doesn't work as intended since it shows cross talk between unrelated workspaces.

I have created a ticket to follow-up in the discussion https://github.com/ros2/ros2/issues/1021.
One of the subitems of that ticket is revisiting the usage of `NO_CMAKE_PACKAGE_REGISTRY` here if we later decide to do something different.

> I just raised the concern that the actual question how we should handle this hasn't been answered yet

I think I have answered how I think it should be handled in https://github.com/ros2/yaml_cpp_vendor/pull/16#discussion_r477600329, https://github.com/ros2/yaml_cpp_vendor/pull/16#issuecomment-686753998.
I think that globally disabling the package registry might not be an option to everybody, and locally disabling it in the places it's known to cause problem seems correct.

I understand that we would like to solve the problem everywhere, but I think that a clear action item for the original author was missing.
I think that asking a contributor to check every vendor package in the project when they are proposing an enhancement in one is a bit to much (though I opened this PR, @seanyen is the author).

> Please feel free to merge before that is the case if you think that is fine.

I think I have satisfied the only [missing request](https://github.com/ros2/yaml_cpp_vendor/pull/16#issuecomment-686716017) of opening a ticket to follow-up in the `NO_CMAKE_PACKAGE_REGISTRY` discussion and listing every package where this is a problem, so I would like to either get an approval or a comment stating what else is missing.

**Review by dirk-thomas — APPROVED** [review]
Repeating my previous comment:

> I just raised the concern that the actual question how we should handle this hasn't been answered yet. Please feel free to merge before that is the case if you think that is fine.

**Comment by ivanpauno** [comment]
> I just raised the concern that the actual question how we should handle this hasn't been answered yet. Please feel free to merge before that is the case if you think that is fine.

:+1: we can follow-up in how we want to handle this in https://github.com/ros2/ros2/issues/1021.

**Comment by ivanpauno** [comment]
Thanks @seanyen !

**Comment by jacobperron** [comment]
@ivanpauno This ticket is on the Foxy and Dashing release boards, can you take care of backporting this change (if applicable). Thanks!

**Comment by ivanpauno** [comment]
> @ivanpauno This ticket is on the Foxy and Dashing release boards, can you take care of backporting this change (if applicable). Thanks!

I don't think we need to backport this.
It was only an improvement to use the system yaml-cpp if available.

**Pull request merged**
